compiler/rustc_hir_analysis/src/check/wfcheck.rs

```diff
@@ -1755,20 +1755,41 @@ fn check_variances_for_type_defn<'tcx>(
             .collect::<FxHashSet<_>>()
     });
 
+    let ty_generics = tcx.generics_of(item.owner_id);
+
     for (index, _) in variances.iter().enumerate() {
         let parameter = Parameter(index as u32);
 
         if constrained_parameters.contains(&parameter) {
             continue;
         }
 
-        let param = &hir_generics.params[index];
+        let ty_param = &ty_generics.params[index];
+        let mut hir_param = &hir_generics.params[index];
+
+        if ty_param.name != hir_param.name.ident().name {
+            // valid programs always have lifetimes before types in the generic parameter list
+            // ty_generics are normalized to be in this required order, and variances are built
+            // from ty generics, not from hir generics. but we need hir generics to get
+            // a span out
+            //
+            // if they aren't in the same order, then the user has written invalid code, and already
+            // got an error about it (or I'm wrong about this)
+            tcx.sess
+                .delay_span_bug(hir_param.span, "hir generics and ty generics in different order");
+            for hp in hir_generics.params {
+                if hp.name.ident().name == ty_param.name {
+                    hir_param = hp;
+                    break;
+                }
+            }
```

Comment:
Should we just silence the error? The program is invalid anyway, and we check that by delaying a bug.
```suggestion
            continue;
```